File: aizu/alds1_7_B.cpp
#include<iostream>
using namespace std;
#define MAX 30
#define NIL -1

struct Node {
  int parent, left, right;
};

Node T[MAX];
int D[MAX], H[MAX], S[MAX];

void setSibling(int u) {
  if (T[u].parent == NIL) {
    S[u] = -1;
  } else {
    if (T[T[u].parent].left != u && T[T[u].parent].left != NIL) {
      S[u] = T[T[u].parent].left;
    } else if (T[T[u].parent].right != u && T[T[u].parent].right != NIL) {
      S[u] = T[T[u].parent].right;
    } else {
      S[u] = -1;
    }
  }
}

void setDepth(int u, int d) {
  // これでuにNILが来ても安全！
  if (u == NIL) {
    return;
  }
  D[u] = d;
  // leftとrightの子どもたちは自分の深さp+1が設定される
  setDepth(T[u].left, d+1);
  setDepth(T[u].right, d+1);
}

int getHeight(int u) {
  int h1 = 0, h2 = 0;
  // 左に子がいる時
  if (T[u].left != NIL) {
    h1 = getHeight(T[u].left) + 1;
  }
  // 右に子がいる時
  if (T[u].right != NIL) {
    h2 = getHeight(T[u].right) + 1;
  }
  if (h1 > h2) {
    return H[u] = h1;
  } else {
    return H[u] = h2;
  }
}

void printNode(int u) {
  // ここで次数を計算する
  int degree = 0;
  if (T[u].left != NIL) {
    degree++;
  }
  if (T[u].right != NIL) {
    degree++;
  }
  cout << "node " << u << ": ";
  cout << "parent = " << T[u].parent << ", ";
  cout << "sibling = " << S[u] << ", ";
  cout << "degree = " << degree << ", ";
  cout << "depth = " << D[u] << ", ";
  cout << "height = " << H[u] << ", ";

  string type = "";
  if (T[u].parent == NIL) {
    type = "root";
  } else if (T[u].left != NIL || T[u].right != NIL) {
    type = "internal node";
  } else {
    type = "leaf";
  }
  cout << type << endl;
}

int main() {

  int n;
  cin >> n;
  // 初期化
  for (int i = 0; i < n; i++) {
    T[i].parent = T[i].left = T[i].right = NIL;
  }
  // 一行ずつ読み込む
  for (int i = 0; i < n; i++) {
    int v, left, right;
    cin >> v >> left >> right;
    // とりあえず左から右に受け流す〜
    // -1が入る可能性があるが、それはNIL扱い
    if (left != NIL) {
      T[v].left = left;
      T[left].parent = v;
    }
    if (right != NIL) {
      T[v].right = right;
      T[right].parent = v;
    }
  }

  int r;
  for (int i = 0; i < n; i++) {
    setSibling(i);
    if (T[i].parent == NIL) {
      r = i;
    }
  }

  setDepth(r, 0);
  getHeight(r);

  for (int i = 0; i < n; i++) {
    printNode(i);
  }

}

File: aizu/alds1_7_B.py
class Node:
    def __init__(self, val):
        self.val = val
        self.parent = -1
        self.left = -1
        self.right = -1
    
    def get_depth(self):
        if self.parent == -1:
            return 0
        else:
            return nodes[self.parent].get_depth() + 1

    def get_degree(self):
        degree = 0
        if self.left != -1:
            degree += 1
        if self.right != -1:
            degree += 1
        return degree

    def get_height(self):
        # if self.parent != -1:
        #     return 0
        max_height = 0
        if self.left != -1:
            max_height = nodes[self.left].get_height() + 1
            # print("left", max_height)
        if self.right != -1:
            max_height = max(max_height, nodes[self.right].get_height() + 1)
            # print("right", max_height)
        return max_height
    
    def get_sibling(self):
        if self.parent == -1:
            return -1
        
        if nodes[self.parent].left == self.val:
            return nodes[self.parent].right
        else:
            return nodes[self.parent].left

    def get_type(self):
        if self.parent == -1:
            return 'root'
        elif self.left == -1 and self.right == -1:
            return 'leaf'
        else:
            return 'internal node'

n = int(input())
nodes = []
for i in range(n):
    nodes.append(Node(i))


for i in range(n):
    # node
    i, left, right = map(int, input().split())
    nodes[i].left = left
    nodes[i].right = right

    # parent
    if left != -1:
        nodes[left].parent = i
    if right != -1:
        nodes[right].parent = i

    # if i == 4:
    #     print("----------")
    #     print(left, nodes[left].parent, right, nodes[right].parent)
    #     print("----------")
    # for j in range(n):
    #     if j == 8:
    #         print('----------')
    #         print(i, nodes[j].parent)x
    #         print('----------')


# 出力
for i in range(n):
    print('node {val}: parent = {p}, sibling = {s}, degree = {degree}, depth = {depth}, height = {height}, {type}'.format(
        val=i,
        p=nodes[i].parent,
        s=nodes[i].get_sibling(),
        degree=nodes[i].get_degree(),
        depth=nodes[i].get_depth(),
        height=nodes[i].get_height(),
        type=nodes[i].get_type(),
    ))